ProjFS.Mac/PrjFSKext/PrjFSKext/KauthHandler.cpp

```diff
@@ -382,28 +384,66 @@ static int HandleFileOpOperation(
         // arg1 is the (const char *) path
         int closeFlags = static_cast<int>(arg2);
         
-        if (KAUTH_FILEOP_CLOSE_MODIFIED == closeFlags)
+        vtype vnodeType = vnode_vtype(currentVnode);
+        if (ShouldIgnoreVnodeType(vnodeType, currentVnode))
         {
-            VirtualizationRoot* root = nullptr;
-            int pid;
-            if (!ShouldHandleFileOpEvent(
-                    context,
+            goto CleanupAndReturn;
+        }
+        
+        if (vnode_isdir(currentVnode))
+        {
+            goto CleanupAndReturn;
+        }
+
+        bool fileFlaggedInRoot = FileIsFlaggedAsInRoot(currentVnode, context);
+        if (fileFlaggedInRoot && KAUTH_FILEOP_CLOSE_MODIFIED != closeFlags)
+        {
+            goto CleanupAndReturn;
+        }
+		
+        if (!fileFlaggedInRoot && !HasAncestorFlaggedAsInRoot(currentVnode, context))
```

Comment:
Do you actually need this call? The next call to `ShouldHandleFileOpEvent` already calls `VirtualizationRoots_FindForVnode` and will return a root if one exists. So it seems like `ShouldHandleFileOpEvent` already returns false if there is no ancestor that is a root.

Reply:
If I understand the design direction correctly (@pmj ?), we will be getting rid of our file flags at some point and our vnode cache will be the source of truth for whether something is a root or not. So I don't think we should spread the implementation details of what it means to be a root. At a minimum, this lock-free approach should be moved into `VirtualizationRoots.hpp/cpp`, but yea it could also be a premature optimization if you haven't actually measured a perf hit.

Reply:
Ok thanks, in that case I'll just take this out for now and rely on `ShouldHandleFileOpEvent`